### PR TITLE
Multiplayer menu

### PR DESCRIPTION
**Pull request by nesrak1**
* Can add/delete servers in the main menu
* Set username before join

### CODE REVIEW

File: NitroxClient/MonoBehaviours/MultiplayerButton.cs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+using System.IO;
+using System.Linq;
+using System.Text;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+using UnityEngine.UI;
+using UWE;
+using NitroxClient.Communication;
+
+namespace NitroxClient.MonoBehaviours
```
Comment (Enflorence) [inline NitroxClient/MonoBehaviours/MultiplayerButton.cs]: I recommend moving classes MainMenuMultiplayerPanel and JoinServerScript into their own files to make it easier to maintain down the road.

File: NitroxClient/MonoBehaviours/Multiplayer.cs [hunk 2]
```diff
@@ -17,6 +17,7 @@ public class Multiplayer : MonoBehaviour
 
         public static Logic Logic;
         public static PacketSender PacketSender; //TODO: migrate logic out of and remove as member variable
+        public static Multiplayer main;
```
Comment (Enflorence) [inline NitroxClient/MonoBehaviours/Multiplayer.cs]: Is there any other way to reference the Multiplayer component without this static reference? At the very least, I would recommend leveraging a static method to grab an instance.

File: NitroxClient/MonoBehaviours/MultiplayerButton.cs (hunk 1 above)
Comment (Sunrunner37) [inline NitroxClient/MonoBehaviours/MultiplayerButton.cs]: Yeah maybe put all of this stuff in a new namespace NitroxClient.MonoBehaviours.Gui.MainMenu

File: NitroxClient/MonoBehaviours/Multiplayer.cs (hunk 2 above)
Comment (nesrak1) [inline NitroxClient/MonoBehaviours/Multiplayer.cs]: We had a discussion about this and I didn't really know what to do. The resolution was that subnautica already uses plenty of singletons, but I do agree it is a bit silly.

File: NitroxClient/MonoBehaviours/MultiplayerButton.cs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        public void Awake()
+        {
+            MultiplayerMenuMods();
+        }
+
+        private void MultiplayerMenuMods()
+        {
+            GameObject startButton = GameObject.Find("Menu canvas/Panel/MainMenu/PrimaryOptions/MenuButtons/ButtonPlay");
+            GameObject showLoadedMultiplayer = Instantiate(startButton);
+            Text buttonText = showLoadedMultiplayer.transform.Find("Circle/Bar/Text").gameObject.GetComponent<Text>() as Text;
```
Comment (Sunrunner37) [inline NitroxClient/MonoBehaviours/MultiplayerButton.cs]: minor: do we need to say "as Text" - it looks like you are already using the generic method call in get component;

File: NitroxClient/MonoBehaviours/MultiplayerButton.cs [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+                HideNewServer();
+            }
+            if (GUI.Button(new Rect(Screen.width / 2 - 250, Screen.height / 2 + 100, 500, 50), "Cancel"))
+            {
+                HideNewServer();
+            }
+        }
+    }
+
+    public class JoinServerScript : MonoBehaviour
```
Comment (Sunrunner37) [inline NitroxClient/MonoBehaviours/MultiplayerButton.cs]: minor: would recommend dropping the word Script as it is implied by the monobehaviour namespace.  Maybe brainstorm a more descriptive postfix(Gui maybe?)

File: NitroxClient/MonoBehaviours/MultiplayerButton.cs [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        }
+
+        public void OnGUI()
+        {
+            if (!showingUsername)
+            {
+                return;
+            }
+            username = GUI.TextField(new Rect(Screen.width / 2 - 250, Screen.height / 2 - 25, 500, 50), username);
+            if (GUI.Button(new Rect(Screen.width / 2 - 250, Screen.height / 2 + 25, 500, 50), "Add server"))
```
Comment (Sunrunner37) [inline NitroxClient/MonoBehaviours/MultiplayerButton.cs]: Note: for the future we'll need to research how they accomplish localization.  I'll add an entry into the Trello area.

File: NitroxPatcher/Patches/uGUI_MainMenu_Start_Patch.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+namespace NitroxPatcher.Patches
+{
+    public class uGUI_MainMenu_Start_Patch : NitroxPatch
+    {
+        public static readonly Type TARGET_CLASS = typeof(uGUI_MainMenu);
+        public static readonly MethodInfo TARGET_METHOD = TARGET_CLASS.GetMethod("Start", BindingFlags.NonPublic | BindingFlags.Instance);
+
+        public static void Postfix(uGUI_MainMenu __instance)
+        {
+            new GameObject().AddComponent<MultiplayerButton>();
```
Comment (Sunrunner37) [inline NitroxPatcher/Patches/uGUI_MainMenu_Start_Patch.cs]: should we use uGui_MainMenu's gameObject instead of creating a new one?

Comment (nesrak1) [inline NitroxPatcher/Patches/uGUI_MainMenu_Start_Patch.cs]: Sure, why not

**Pull request merged**